src/query/execute_statement.c

```diff
@@ -14532,118 +14491,155 @@ do_prepare_session_statement (PARSER_CONTEXT * parser, PT_NODE * statement)
  * do_prepare_cte () - prepare CTE statements in WITH clause for query cache
  * return : Error code
  * parser (in)	  : parser context
- * statement (in) : statement to prepare
+ * stmt (in) : statement to prepare
  */
-static int
-do_prepare_cte (PARSER_CONTEXT * parser, PT_NODE * statement)
+int
+do_prepare_cte (PARSER_CONTEXT * parser, PT_NODE * stmt)
 {
-  int error = NO_ERROR;
-  PT_NODE *cte_statement;
-  PT_NODE *cte_def_list;
+  int err = NO_ERROR;
   PARSER_CONTEXT cte_context;
-  int var_count;
+  int i, var_count;
+
+  cte_context = *parser;
 
-  for (cte_def_list = statement; cte_def_list; cte_def_list = cte_def_list->next)
+  cte_context.host_var_count = 0;
+  var_count = parser->host_var_count + parser->auto_param_count;
+
+  stmt->info.query.flag.cte_query_cached = 1;
+
+  if (var_count > 0)
     {
-      if (cte_def_list->info.cte.non_recursive_part->info.query.hint & PT_HINT_QUERY_CACHE)
+      cte_context.cte_host_var_index = (int *) parser_alloc (parser, var_count * sizeof (int));
+      if (cte_context.cte_host_var_index == NULL)
 	{
-	  cte_statement = cte_def_list->info.cte.non_recursive_part;
+	  er_set (ER_ERROR_SEVERITY, ARG_FILE_LINE, ER_OUT_OF_VIRTUAL_MEMORY, 1, var_count * sizeof (int));
+	  return ER_OUT_OF_VIRTUAL_MEMORY;
+	}
+      /* to ininitialize empty index */
+      memset ((void *) cte_context.cte_host_var_index, -1, var_count * sizeof (int));
+    }
 
-	  cte_context = *parser;
-	  cte_context.dbval_cnt = 0;
-	  cte_context.host_var_count = cte_context.auto_param_count = 0;
+  parser_walk_tree (&cte_context, stmt, pt_cte_host_vars_index, parser, NULL, NULL);
```

Comment:
pt_cte_host_vars_index() has been changed so that the argument is no longer needed.
```suggestion
  parser_walk_tree (&cte_context, stmt, pt_cte_host_vars_index, NULL, NULL, NULL);
```

Reply:
good point. thanks.

Reply:
it needs the context to get host variable indext from original parser.